### PR TITLE
Docs 0.2

### PR DESCRIPTION
**Pull request by numerator**
### Description
Review notes: Please review in 2 ways:
- standard code review, mainly focusing on the changes noted below to the typedoc generation and integration into docusaurus
- content review, by pulling the branch and doing:
```
$ yarn docs
$ cd docs
$ yarn start
```
This will generate the typedoc, then will launch the docusaurus site in your browser. Please poke around to make sure no broken links and also read the new content to make sure it makes sense.

- Completed docs for 0.2, including
- Added a *guide* for JustInLite on GCP
- Added a "concepts" page for Logging and Result Writing
- Changed "using" and "concepts/overview" pages to note differences between JustIn "standard" and JustInLite, and directed readers to the correct content.
- Significantly changed (with the help of AI), the way that typedoc is generated so that
    - it only generates docs for JustInWrapper, JustInLiteWrapper, and the types referenced by those
    - docs are generated in markdown instead of HTML and written into the docusaurus file hiearchy under "/api"
    - a script (./scripts/generate-api-sidebar.js) runs after typedoc generation to generate a cleaned-up sidebar nav structure for the typedoc (./docs/api-sidebar.js), which is then referenced in sidebars.ts to integrate the api nav into the overall docs nav.

NOTE: The handler types documentation is a mess and very hard to follow because of all the TypeScript mojo. To discuss: should we clean this up so that the docs are easier to follow?

### Related Jira Tickets

-

### Comments 

-

Please ensure all items are complete before requesting a review:

- [ ] **Code Completed**: All code for the feature or bugfix has been implemented.
- [ ] **Unit Tests Added**: Unit tests are written for all new functionality, covering edge cases where applicable.
- [ ] **Integration Tests (where necessary)**: Integration tests are added for interactions between modules or components.
- [ ] **All Tests Passing**: All unit and integration tests are passing.
- [ ] **Type Safety Checked**: Ensure TypeScript types are correctly defined and used. Avoid `any` types without proper justification.
- [ ] **Typedoc Comments Added**: Ensure all functions, classes, interfaces, and types are documented using TypeDoc syntax.
- [ ] **Code Formatting**: All code files are formatted according to the project’s standards (e.g., Prettier, ESLint) and no linting errors remain.
- [ ] **No Console Logs**: Remove any `console.log` or other debugging output from the code.
- [ ] **Documentation Updated**: Update relevant project documentation (e.g., README, configuration guides) if necessary.

---

### CODE REVIEW

**Comment by jadrubio** [comment]
couple of comments to put in the description.  you will want to make sure you run yarn install in the docs directory incase you dont have the proper modules.  Also when you look at the site locally change the version to Next

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@just-in/core",
-  "version": "0.1.0",
+  "version": "0.2.0",
```
Comment (jadrubio) [inline package.json:3]: this is a little concerning since main should already be at 0.2.0

File: docs/sidebars.ts [hunk 2]
```diff
@@ -11,21 +11,26 @@ const sidebars: SidebarsConfig = {
         {type: 'doc', id: 'concepts/handlers'},
         {type: 'doc', id: 'concepts/events'},
         {type: 'doc', id: 'concepts/users'},
+        {type: 'doc', id: 'concepts/logging'}
       ],
     },
     {
       type: 'category',
-      label: 'API Reference',
+      label: 'Guides',
```
Comment (jadrubio) [inline docs/sidebars.ts:19]: I think Guides should be at the bottom of the sidebars as they talk about non JustIn things

File: docs/docusaurus.config.ts [hunk 3]
```diff
@@ -35,7 +35,7 @@ const config: Config = {
           sidebarPath: require.resolve('./sidebars.ts'),
           showLastUpdateAuthor: true,
           showLastUpdateTime: true,
-          includeCurrentVersion: false,
+          includeCurrentVersion: true,
```
Comment (jadrubio) [inline docs/docusaurus.config.ts:38]: is this something we will be turning back off?  I think we had discussions on whether or not to show the Next docs

File: docs/docs/using.md [hunk 4]
```diff
@@ -4,35 +4,42 @@ title: Using JustIn Core
 sidebar_position: 1
 ---
 
-
 # Justin Core
```
Comment (jadrubio) [inline docs/docs/using.md:7]: We need to standardized Justin.  some times we have Justin, sometimes we have justin, other times we have JustIn, the package is just-in ....

we should knuckle down on what we want to use

File: docs/docs/using.md [hunk 5]
```diff
@@ -4,35 +4,42 @@ title: Using JustIn Core
...
 
 The `@just-in/core` package is a lightweight, **event-driven** framework for building **Just-In-Time Adaptive Interventions (JITAIs)**. It helps research app developers define, schedule, and deliver adaptive interventions using events, decision rules, and tasks.
```
Comment (jadrubio) [inline docs/docs/using.md:9]: out of the scope of this PR but do we want to have events, decision rules, tasks link to their respective page in the docs?  We can save this change for the events-engine docs

File: docs/docs/using.md [hunk 6]
```diff
@@ -4,35 +4,42 @@ title: Using JustIn Core
...
-```bash
-$ mkdir my-justin-app
-$ cd my-justin-app
-$ npm init 
-$ npm install @just-in/core
-```
-
-You can also start with one of our [examples](https://github.com/MIACollaborative/justin-examples) (recommended), or our tutorial.
+There are two ways to run `@just-in/core`, **standard mode** and **serverless mode** (a.k.a. "JustInLite"). 
+* **Standard mode** requires a connection to a MongoDB server. We have [guides to walk you through selecting and configuring a MongoDB installation](./guides/mongo.md) to work with JustIn.
```
Comment (jadrubio) [inline docs/docs/using.md:18]: I think we should say it requires a connection to a db - currently we only support mongodb.


also i like the full justin but i can see why you wouldnt use that here 😅

File: docs/docs/using.md [hunk 7]
```diff
@@ -4,35 +4,42 @@ title: Using JustIn Core
...
-$ mkdir my-justin-app
-$ cd my-justin-app
-$ npm init 
-$ npm install @just-in/core
-```
-
-You can also start with one of our [examples](https://github.com/MIACollaborative/justin-examples) (recommended), or our tutorial.
+There are two ways to run `@just-in/core`, **standard mode** and **serverless mode** (a.k.a. "JustInLite"). 
+* **Standard mode** requires a connection to a MongoDB server. We have [guides to walk you through selecting and configuring a MongoDB installation](./guides/mongo.md) to work with JustIn. 
+* **Serverless** mode is designed to be run as a cloud function/lambda within a cloud services environment such as Google Cloud or Amazon Web Services. See the [Serverless with JustInLite and Google Cloud guide](./guides/gcp.md) for instructions on setting up JustInLite.
```
Comment (jadrubio) [inline docs/docs/using.md:19]: Being pedantic but do we want to say to be ran without a dedicated db typically as a cloud function or lambda etc etc?

File: docs/docs/guides/gcp.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
+---
```
Comment (jadrubio) [inline docs/docs/guides/gcp.md:1]: I am wondering if these guides should be on the main justin-framework site or dispersed on the other sites.  I am sure that some of this may be pertinent to two different packages or may not be logical for a 3PD to look at a different packages docs for this information

File: docs/docs/guides/gcp.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+import sched6 from '/img/sched6-options.png';
+import sched7 from '/img/sched7-list-paused.png';
+import sched8 from '/img/sched8-list-running.png';
+import run1 from '/img/run1-logs.png';
+
+
+# Serverless Execution with JustInLite and Google Cloud
+Modern cloud computing platforms like Google Cloud Platform and Amazon Web Services offer tools for "serverless execution" which allow developers to run application code without needing to set up and maintain servers, databases, and the like. This guide will show you how to execute your Decision Rules and Tasks using features of Google Cloud Platform, including Cloud Scheduler and Cloud Run Functions. 
+
+## JustInLiteWrapper
```
Comment (jadrubio) [inline docs/docs/guides/gcp.md:20]: I think this is an under the hood name and may not need to be displayed here at least not as a single word

File: docs/docs/guides/gcp.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+import sched4 from '/img/sched4-create-topic.png';
+import sched5 from '/img/sched5-topic-body.png';
+import sched6 from '/img/sched6-options.png';
+import sched7 from '/img/sched7-list-paused.png';
+import sched8 from '/img/sched8-list-running.png';
+import run1 from '/img/run1-logs.png';
+
+
+# Serverless Execution with JustInLite and Google Cloud
+Modern cloud computing platforms like Google Cloud Platform and Amazon Web Services offer tools for "serverless execution" which allow developers to run application code without needing to set up and maintain servers, databases, and the like. This guide will show you how to execute your Decision Rules and Tasks using features of Google Cloud Platform, including Cloud Scheduler and Cloud Run Functions.
```
Comment (jadrubio) [inline docs/docs/guides/gcp.md:18]: I would make a caveat that this is a quick and dirty tutorial and we recommend looking at the latest docs for what ever cloud platform you are using etc etc

File: docs/docs/guides/gcp.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+1. It does not require a connection to a Mongo database.
+2. It does not create or use event generators such as the built-in `IntervalTimerEventGenerator` to create and publish events.
+
+Instead, `JustInLite` apps depend on external services for reading and writing data, and leverage the host cloud environment for generating events that trigger rule execution.
+
+Importantly, the lifecycle of a `JustInLite` app is *very different* from that of a typical `JustIn` app. Whereas a typical `JustIn` app is expected to run indefinitely once started, a `JustInLite` app is short-lived. The `JustInLite` app lifecycle is as follows:
+* An external event (e.g., a Cloud Scheduler event) triggers the invocation of the app (e.g., by invoking a Cloud Run Function that contains the app code).
+* The `JustInLite` app calls external APIs to assemble the resources it needs (e.g., user records, content objects) for its `DecisionRule`s and `Task`s. The app calls `JustInLiteWrapper.loadUsers()` with the user records obtained from the external service.
```
Comment (jadrubio) [inline docs/docs/guides/gcp.md:29]: I think this is misleading - JustIn Lite does not make the call to 3rd party apis - the 3PD will make the call and transform the user to make sure it meets our JUser or NewUserRecord shape and then load it into Justin.

I would shorten up the logging bullet point and move it up noting it will default to the console but can be overridden.

Then i would discuss the loading of the users, then the registration of the tasks and decision rules, then the registration of events, potential for branching logic to kick off some events vs others etc etc, the how to kick off an event.


actually all this should be in the Justin-Lite section I would just link from here to there to see the flow

**Review by peiyaoh — COMMENTED** [review]
Hi @numerator, I compiled my suggestions/comments here.

1. Clarify/define what we mean by external services (which includes services provided by a cloud environment). I think saying getting data from external services is potentially confusing. For instance, the following statement might create some uncertainty (or confusion) about what additional services are needed for JustInLite to work.
> The app calls `JustInLiteWrapper.loadUsers()` with the user records obtained from the external service.

2. It could be beneficial to categorize the content of JustInLiteWrapper into sub-sections to highlight different steps in the lifecycle, such as
	- Ping from the cloud environment (or external services)
	- Load data: loadUsers
	- Configure system: log configuration, rule/task registration, registerEventHandlers
	- Execute rule/task: publishEvent, (rule and task execution)

3.  Consider explaining the rest of the parameters of the gcloud command, at least at the high level (for those that are important), as there are a lot of parameters. Pointing to documentation could also be great.
4. In JustInLiteWrapper, consider explaining "warm instance" or providing a reference/link for the definition, as we are targeting a CS undergrad and they might not understand what it means.
